File: algorithm/src/main/java/baekjoon/p02578/Main.java
package baekjoon.p02578;

import java.util.Scanner;

public class Main {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int[][] a = new int[5][5];
        int[] b = new int[25];

        for (int i = 0; i < a.length; i++) {
            for (int j = 0; j < a[i].length; j++) {
                a[i][j] = sc.nextInt();
            }
        }

        for (int i = 0; i < b.length; i++) {
            b[i] = sc.nextInt();
        }

        System.out.println(solve(a, b));
    }

    private static int solve(int[][] a, int[] b) {
        int check = -1;

        for (int i = 0; i < b.length; i++) {
            check(a, b[i], check);
            if (bingo(a, check) >= 3) {
                return i + 1;
            }
        }

        return -1;
    }

    private static void check(int[][] arr, int target, int check) {
        for (int i = 0; i < arr.length; i++) {
            for (int j = 0; j < arr[i].length; j++) {
                if (arr[i][j] == target) {
                    arr[i][j] = check;
                    break;
                }
            }
        }
    }

    private static int bingo(int[][] arr, int check) {
        int ret = 0;
        int len = arr.length;

        for (int i = 0; i < len; i++) {
            if (all(arr[i], check)) {
                ret++;
            }
        }

        for (int i = 0; i < len; i++) {
            int[] temp = new int[len];
            for (int j = 0; j < len; j++) {
                temp[j] = arr[j][i];
            }

            if (all(temp, check)) {
                ret++;
            }
        }

        int[] a = new int[len];
        for (int i = 0; i < len; i++) {
            a[i] = arr[i][i];
        }

        int[] b = new int[len];
        for (int i = 0; i < len; i++) {
            b[i] = arr[i][len - i - 1];
        }

        if (all(a, check)) {
            ret++;
        }

        if (all(b, check)) {
            ret++;
        }

        return ret;
    }

    private static boolean all(int[] arr, int target) {
        for (int e : arr) {
            if (e != target) {
                return false;
            }
        }

        return true;
    }

}

File: algorithm/src/main/java/baekjoon/p02578/Main.kt
package baekjoon.p02578

fun main(args: Array<String>) {
    val a = List(5) { readLine()!!.split(" ").map { it.toInt() }.toIntArray() }
    val b = List(5) { readLine()!!.split(" ").map { it.toInt() } }
    println(solve(a, b))
}

private fun solve(a: List<IntArray>, b: List<List<Int>>): Int {
    val check = 0
    val size = b.size

    for (i in 0 until size) {
        for (j in 0 until size) {
            check(a, b[i][j], check)
            if (bingo(a, check) >= 3) {
                return (i * size + j + 1)
            }
        }
    }

    return -1
}

private fun check(arr: List<IntArray>, target: Int, check: Int) {
    for (i in 0 until arr.size) {
        for (j in 0 until arr[i].size) {
            if (arr[i][j] == target) {
                arr[i][j] = check
            }
        }
    }
}

private fun bingo(arr: List<IntArray>, check: Int): Int {
    var ret = 0
    val size = arr.size

    for (e in arr) {
        if (e.all { it == check }) {
            ret++
        }
    }

    for (i in 0 until size) {
        val temp = IntArray(size)
        for (j in 0 until size) {
            temp[j] = arr[j][i]
        }

        if (temp.all { it == check }) {
            ret++
        }
    }

    val a = IntArray(size)
    val b = IntArray(size)

    for (i in 0 until size) {
        a[i] = arr[i][i]
    }

    for (i in 0 until size) {
        b[i] = arr[i][size - i - 1]
    }

    if (a.all { it == 0 }) ret++
    if (b.all { it == 0 }) ret++

    return ret
}